<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Pedigree kinship() details • Pedixplorer</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Pedigree kinship() details">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Pedixplorer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/pedigree_alignment.html">Pedigree alignment details</a>
    </li>
    <li>
      <a href="../articles/pedigree_kinship.html">Pedigree kinship() details</a>
    </li>
    <li>
      <a href="../articles/pedigree_object.html">Pedigree object</a>
    </li>
    <li>
      <a href="../articles/pedigree_plot.html">Pedigree plotting details</a>
    </li>
    <li>
      <a href="../articles/pedigree.html">Pedixplorer tutorial</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/LouisLeNezet/Pedixplorer/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Pedigree kinship() details</h1>
                        <h4 data-toc-skip class="author">TM
Therneau</h4>
            
            <h4 data-toc-skip class="date">03 October, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/LouisLeNezet/Pedixplorer/blob/HEAD/vignettes/pedigree_kinship.Rmd" class="external-link"><code>vignettes/pedigree_kinship.Rmd</code></a></small>
      <div class="hidden name"><code>pedigree_kinship.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The kinship matrix is foundational for random effects models with
family data. For
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>
subjects it is an
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi><mo>×</mo><mi>n</mi></mrow><annotation encoding="application/x-tex">n \times n</annotation></semantics></math>
matrix whose
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
elements contains the expected fraction of alleles that would be
identical by descent if we sampled one from subject
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
and another from subject
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>.
Note that the diagonal elements of the matrix will be 0.5 not 1: when we
randomly sample twice from the same subject (with replacement) we will
get two copies of the gene inherited from the father 1/4 of the time,
the maternal copy twice (1/4) or one of each 1/2 the time. The formal
definition is
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo>,</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mn>1</mn><mi>/</mi><mn>4</mn><mo>+</mo><mn>1</mn><mi>/</mi><mn>4</mn><mo>+</mo><mn>1</mn><mi>/</mi><mn>2</mn><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>m</mi><mo>,</mo><mi>f</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">K(i,i) = 1/4 + 1/4 + 1/2 K(m,f)</annotation></semantics></math>
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>m</mi><annotation encoding="application/x-tex">m</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>f</mi><annotation encoding="application/x-tex">f</annotation></semantics></math>
are the father and mother of subject
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>.</p>
<p>The algorithm used is found in K Lange, <em>Mathematical and
Statistical Methods for Genetic Analysis</em>, Springer 1997, page
71–72.</p>
<p>The key idea of the recursive algorithm for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo>,</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">K(i,j)</annotation></semantics></math>
is to condition on the gene selection for the first index
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>.
Let
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>m</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">m(i)</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>f</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">f(i)</annotation></semantics></math>
be the indices of the mother and father of subject
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>
be the allele randomly sampled from subject
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>,
which may of either maternal or paternal origin.</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo>,</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>g</mi><mrow><mspace width="0.333em"></mspace><mtext mathvariant="normal"> maternal</mtext></mrow><mo stretchy="true" form="postfix">)</mo></mrow><mo>*</mo><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>m</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>+</mo><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>g</mi><mrow><mspace width="0.333em"></mspace><mtext mathvariant="normal"> paternal</mtext></mrow><mo stretchy="true" form="postfix">)</mo></mrow><mo>*</mo><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>f</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mn>1</mn><mi>/</mi><mn>2</mn><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>m</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>+</mo><mn>1</mn><mi>/</mi><mn>2</mn><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>f</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo>,</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mn>1</mn><mi>/</mi><mn>2</mn><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>+</mo><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>m</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo><mi>f</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow></mtd></mtr></mtable><annotation encoding="application/x-tex">
\begin{align}
    K(i,j) &amp;= P(\mbox{$g$ maternal}) * K(m(i), j) + 
                P(\mbox{$g$ paternal}) * K(f(i), j) \label{recur0} \\
            &amp;= 1/2 K(m(i), j) + 1/2 K(f(i), j)   \label{recur1} \\
    K(i,i) &amp;= 1/2(1 + K(m(i), f(i))) \label{self} 
\end{align}
</annotation></semantics></math></p>
<p>The key step in equation
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo>,</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">K(i,j)</annotation></semantics></math>
is if
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>
has a maternal origin, then it is a random selection from the two
maternal genes, and its IBD state with respect to subject
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
is that of a random selection from
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>m</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">m(i)</annotation></semantics></math>
to a random selection from
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>.
This is precisely the definition of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>m</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">K(m(i), j)</annotation></semantics></math>.
The recursion does not work for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo>,</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">K(i,i)</annotation></semantics></math>
since once we select a maternal gene the second choice from
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
cannot use a different maternal gene.</p>
<p>For the recurrence algorithm to work properly we need to compute the
values of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>K</mi><annotation encoding="application/x-tex">K</annotation></semantics></math>
for any parent before the calculations for their children. Pedigree
founders (those with no parents) are assumed to be unassociated, so for
these subjects we have</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo>,</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mn>1</mn><mi>/</mi><mn>2</mn></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo>,</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow></mtd><mtd columnalign="left" style="text-align: left"><mo>=</mo><mn>0</mn><mspace width="0.222em"></mspace><mo>;</mo><mi>i</mi><mo>≠</mo><mi>j</mi></mtd></mtr></mtable><annotation encoding="application/x-tex">
\begin{align*}
    K(i,i) &amp;= 1/2 \\
    K(i,j) &amp;= 0 \ ; i\ne j
\end{align*}
</annotation></semantics></math></p>
<p>The final formula slightly different for the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
chromosome. Equation
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo>,</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">K(i,j)</annotation></semantics></math>
still holds, but for males the probability that a selected
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
chromosome is maternal is 1, so when
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
a male the recurrence formula becomes
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo>,</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>m</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">K(i,j) = K(m(i),j)</annotation></semantics></math>.<br>
For females it is unchanged. All males will have
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo>,</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">K(i,i) = 1</annotation></semantics></math>
for the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
chromosome.</p>
<p>In order to have already-defined terms on the right hand side of the
recurrence formula for each element, subjects need to be processed in
the following order</p>
<ul>
<li>Generation 0 (founders)</li>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo>,</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">K(i,j)</annotation></semantics></math>
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
is from generation 1 and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
from generation 0.</li>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo>,</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">K(i,j)</annotation></semantics></math>
with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
from generation 1</li>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo>,</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">K(i,j)</annotation></semantics></math>
with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
from generation 2 and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
from generation 0 or 1</li>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo>,</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">K(i,j)</annotation></semantics></math>
with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
from generation 2.</li>
<li>…</li>
</ul>
<p>The kindepth routine assigns a plotting depth to each subject in such
a way that parents are always above children.<br>
For each depth we need to do the compuations of formula
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo>,</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">K(i,j)</annotation></semantics></math>
twice. The first time it will get the relationship between each subject
and prior generations correct, the second will correctly compute the
values between subjects on the same level. The computations within any
stage of the above list can be vectorized, but not those between
stages.</p>
<p>Let
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><mi>n</mi><mi>d</mi><mi>x</mi></mrow><annotation encoding="application/x-tex">indx</annotation></semantics></math>
be the index of the rows for the generation currently being processed,
say generation
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>.<br>
We add correct computations to the matrix one row at a time; all of the
calculations depend only on the prior rows with the exception of the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><mo>,</mo><mi>i</mi></mrow><annotation encoding="application/x-tex">i,i</annotation></semantics></math>
element. This approach leads to a for loop containing operations on
single rows/columns.</p>
<p>At one point below we use a vectorized version. It looks like the
snippet below</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span> <span class="op">(</span><span class="va">g</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">depth</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">indx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">depth</span> <span class="op">==</span> <span class="va">g</span><span class="op">)</span></span>
<span>    <span class="va">kmat</span><span class="op">[</span><span class="va">indx</span>, <span class="op">]</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">kmat</span><span class="op">[</span><span class="va">mother</span><span class="op">[</span><span class="va">indx</span><span class="op">]</span>, <span class="op">]</span> <span class="op">+</span> <span class="va">kmat</span><span class="op">[</span><span class="va">father</span><span class="op">[</span><span class="va">indx</span><span class="op">]</span>, <span class="op">]</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span></span>
<span>    <span class="va">kmat</span><span class="op">[</span>, <span class="va">indx</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">kmat</span><span class="op">[</span>, <span class="va">mother</span><span class="op">[</span><span class="va">indx</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="va">kmat</span><span class="op">[</span>, <span class="va">father</span><span class="op">[</span><span class="va">indx</span><span class="op">]</span>, <span class="op">]</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="va">indx</span><span class="op">)</span> <span class="va">kmat</span><span class="op">[</span><span class="va">j</span>, <span class="va">j</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="va">kmat</span><span class="op">[</span><span class="va">mother</span><span class="op">[</span><span class="va">j</span><span class="op">]</span>, <span class="va">father</span><span class="op">[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>The first line computes all the values for a horizontal stripe of the
matrix. It will be correct for columns in generations
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>&lt;</mo><mi>g</mi></mrow><annotation encoding="application/x-tex">&lt;g</annotation></semantics></math>,
unreliable for generation
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>
with itself because of incomplete parental relationships, and zero for
higher generations. The second line does the vertical stripe, and
because of the line before it does have the data it needs and so gets
all the stripe correct. Except of course for the diagonal elements, for
which formula
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>K</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo>,</mo><mi>j</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">K(i,j)</annotation></semantics></math>
does not hold. We fill those in last. We know that vectorized
calculations are always faster in R and I was excited to figure this
out. The unfortunate truth is that for this code it hardly makes a
difference, and for the X chromosome calculation leads to impenetrable
if-then-else logic.</p>
<p>The program can be called with a Pedigree, or raw data. The first
argument is
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><mi>d</mi></mrow><annotation encoding="application/x-tex">id</annotation></semantics></math>
instead of the more generic
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>x</mi><annotation encoding="application/x-tex">x</annotation></semantics></math>
for backwards compatability with an older version of the routine. We
give founders a fake parent of subject
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi><mo>+</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">n+1</annotation></semantics></math>
who is not related to anybody (even themself); it avoids some
if-then-else constructions.</p>
<div class="section level3">
<h3 id="with-a-pedigree-object">With a Pedigree object<a class="anchor" aria-label="anchor" href="#with-a-pedigree-object"></a>
</h3>
<p>The method for a Pedigree object is an almost trivial modification.
Since the mother and father are already indexed into the id list it has
two lines that are different, those that create mrow and drow. The other
change is that now we potentially have information available on
monozygotic twins. If there are any such, then when the second twin of a
pair is added to the matrix, we need to ensure that the pairs kinship
coefficient is set to the self-self value. This can be done after each
level is complete, but before children for that level are computed. If
there are monozygotic triples, quadruplets, etc. this computation gets
more involved.</p>
<p>The total number of monozygotic twins is always small, so it is
efficient to fix up all the monzygotic twins at each generation. A
variable
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>h</mi><mi>a</mi><mi>v</mi><mi>e</mi><mi>m</mi><mi>z</mi></mrow><annotation encoding="application/x-tex">havemz</annotation></semantics></math>
is set to <code>TRUE</code> if there are any, and an index array
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>m</mi><mi>z</mi><mi>i</mi><mi>n</mi><mi>d</mi><mi>e</mi><mi>x</mi></mrow><annotation encoding="application/x-tex">mzindex</annotation></semantics></math>
is created for matrix subscripting.</p>
</div>
<div class="section level3">
<h3 id="with-multiple-families">With multiple families<a class="anchor" aria-label="anchor" href="#with-multiple-families"></a>
</h3>
<p>For the Minnesota Family Cancer Study there are 461 families and
29114 subjects. The raw kinship matrix would be 29114 by 29114 which is
over 5 terabytes of memory, something that clearly will not work within
S.<br>
The solution is to store the overall matrix as a sparse Matrix object.
Each family forms a single block. For this study we have
<code>n &lt;- table(minnbreast$famid); sum(n*(n+1)/2)</code> or 1.07
million entries; assuming that only the lower half of each matrix is
stored. The actual size is actually smaller than this, since each family
matrix will have zeros in it — founders for instance are not related —
and those zeros are also not stored.</p>
<p>The result of each per-family call to kinship will be a symmetric
matrix. We first turn each of these into a dsCMatrix object, a sparse
symmetric form. The
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>b</mi><mi>d</mi><mi>i</mi><mi>a</mi><mi>g</mi></mrow><annotation encoding="application/x-tex">bdiag</annotation></semantics></math>
function is then used to paste all of these individual sparse matrices
into a single large matrix.</p>
<p>Why do we note use <code>(i in famlist)</code> below? A numeric
subscript of <code>[[9]]</code> % selects the ninth family, not the
family labeled as 9, so a numeric family id would not act as we wished.
If all of the subject ids are unique, across all families, the final
matrix is labeled with the subject id, otherwise it is labeled with
family/subject.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LouisLeNezet/Pedixplorer" class="external-link">Pedixplorer</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">sampleped</span><span class="op">)</span></span>
<span><span class="va">ped</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Pedigree-class.html">Pedigree</a></span><span class="op">(</span><span class="va">sampleped</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/kinship.html">kinship</a></span><span class="op">(</span><span class="va">ped</span><span class="op">)</span><span class="op">[</span><span class="fl">35</span><span class="op">:</span><span class="fl">48</span>, <span class="fl">35</span><span class="op">:</span><span class="fl">48</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## 14 x 14 sparse Matrix of class "dsCMatrix"</span></span></code></pre>
<pre><code><span><span class="co">##   [[ suppressing 14 column names '1_135', '1_136', '1_137' ... ]]</span></span></code></pre>
<pre><code><span><span class="co">##                                                                                </span></span>
<span><span class="co">## 1_135 0.500 .     .    0.25 0.125 0.125 0.125 .    .    .   .    .    .    .   </span></span>
<span><span class="co">## 1_136 .     0.500 .    0.25 0.125 0.125 0.125 .    .    .   .    .    .    .   </span></span>
<span><span class="co">## 1_137 .     .     0.50 .    0.250 0.250 0.250 .    .    .   .    .    .    .   </span></span>
<span><span class="co">## 1_138 0.250 0.250 .    0.50 0.250 0.250 0.250 .    .    .   .    .    .    .   </span></span>
<span><span class="co">## 1_139 0.125 0.125 0.25 0.25 0.500 0.250 0.250 .    .    .   .    .    .    .   </span></span>
<span><span class="co">## 1_140 0.125 0.125 0.25 0.25 0.250 0.500 0.250 .    .    .   .    .    .    .   </span></span>
<span><span class="co">## 1_141 0.125 0.125 0.25 0.25 0.250 0.250 0.500 .    .    .   .    .    .    .   </span></span>
<span><span class="co">## 2_201 .     .     .    .    .     .     .     0.50 .    .   0.25 0.25 0.25 0.25</span></span>
<span><span class="co">## 2_202 .     .     .    .    .     .     .     .    0.50 .   0.25 0.25 0.25 0.25</span></span>
<span><span class="co">## 2_203 .     .     .    .    .     .     .     .    .    0.5 .    .    .    .   </span></span>
<span><span class="co">## 2_204 .     .     .    .    .     .     .     0.25 0.25 .   0.50 0.25 0.25 0.25</span></span>
<span><span class="co">## 2_205 .     .     .    .    .     .     .     0.25 0.25 .   0.25 0.50 0.25 0.25</span></span>
<span><span class="co">## 2_206 .     .     .    .    .     .     .     0.25 0.25 .   0.25 0.25 0.50 0.25</span></span>
<span><span class="co">## 2_207 .     .     .    .    .     .     .     0.25 0.25 .   0.25 0.25 0.25 0.50</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="makekinship">MakeKinship<a class="anchor" aria-label="anchor" href="#makekinship"></a>
</h3>
<p>The older
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>m</mi><mi>a</mi><mi>k</mi><mi>e</mi><mi>k</mi><mi>i</mi><mi>n</mi><mi>s</mi><mi>h</mi><mi>i</mi><mi>p</mi></mrow><annotation encoding="application/x-tex">makekinship</annotation></semantics></math>
function, from before the creation of PedigreeList objects, accepts the
raw identifier data, along with a special family code for unrelated
subjects, as produced by the <code><a href="../reference/make_famid.html">make_famid()</a></code> function. All the
unrelated subjects are put at the front of the kinship matrix in this
case rather than within the family. Because unrelateds get put into a
fake family, we cannot create a rational family/subject identifier; the
id must be unique across families.</p>
<p>We include a copy of the routine for backwards compatability, but do
not anticipate any new usage of it. Like most routines, this starts out
with a collection of error checks.</p>
</div>
<div class="section level3">
<h3 id="monozygotic-twins-for-pedigree-object">Monozygotic twins, for Pedigree object<a class="anchor" aria-label="anchor" href="#monozygotic-twins-for-pedigree-object"></a>
</h3>
<p>Return now to the question of monzygotic sets, used specifically in
the kinship for Pedigree objects. Consider the following rather
difficult example:</p>
<table class="table">
<thead><tr class="header">
<th>id1</th>
<th>id2</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>2</td>
</tr>
<tr class="even">
<td>2</td>
<td>3</td>
</tr>
<tr class="odd">
<td>5</td>
<td>6</td>
</tr>
<tr class="even">
<td>3</td>
<td>7</td>
</tr>
<tr class="odd">
<td>10</td>
<td>9</td>
</tr>
</tbody>
</table>
<p>Subjects 1, 2, 3, and 7 form a monozygotic quadruple, 5/6 and 9/10
are monzygotic pairs.<br>
First create a vector <code>mzgrp</code> which contains for each subject
the lowest index of a monozygotic twin for that subject.<br>
For non-twins it can have any value.<br>
For this example that vector is set to 1 for subjects 1, 2, 3, and 7, to
5 for 5 and 6, and to 9 for 9 and 10. Creating this requires a short
while loop. Once this is in hand we can identify the sets.</p>
<p>Next make a matrix that has a row for every possible pair. Finally,
remove the rows that are identical. The result is a set of all pairs of
observations in the matrix that correspond to monozygotic pairs.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">5</span>, <span class="fl">6</span>, <span class="fl">7</span>, <span class="fl">8</span><span class="op">)</span>,</span>
<span>    dadid <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">4</span>, <span class="cn">NA</span>, <span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">4</span>, <span class="cn">NA</span><span class="op">)</span>,</span>
<span>    momid <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">8</span>, <span class="fl">8</span>, <span class="fl">8</span>, <span class="cn">NA</span>, <span class="fl">8</span>, <span class="fl">8</span>, <span class="fl">8</span>, <span class="cn">NA</span><span class="op">)</span>,</span>
<span>    sex <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">rel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    id1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">6</span>, <span class="fl">7</span><span class="op">)</span>,</span>
<span>    id2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">5</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>    code <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">ped</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Pedigree-class.html">Pedigree</a></span><span class="op">(</span><span class="va">df</span>, <span class="va">rel</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">ped</span><span class="op">)</span></span></code></pre></div>
<p><img src="pedigree_kinship_files/figure-html/kinship%20twins-1.png" width="700"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">twins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">7</span>, <span class="fl">5</span>, <span class="fl">6</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/kinship.html">kinship</a></span><span class="op">(</span><span class="va">ped</span><span class="op">)</span><span class="op">[</span><span class="va">twins</span>, <span class="va">twins</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## 6 x 6 sparse Matrix of class "dsCMatrix"</span></span>
<span><span class="co">##      1    2    3    7    5    6</span></span>
<span><span class="co">## 1 0.50 0.50 0.50 0.50 0.25 0.25</span></span>
<span><span class="co">## 2 0.50 0.50 0.50 0.50 0.25 0.25</span></span>
<span><span class="co">## 3 0.50 0.50 0.50 0.50 0.25 0.25</span></span>
<span><span class="co">## 7 0.50 0.50 0.50 0.50 0.25 0.25</span></span>
<span><span class="co">## 5 0.25 0.25 0.25 0.25 0.50 0.50</span></span>
<span><span class="co">## 6 0.25 0.25 0.25 0.25 0.50 0.50</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: x86_64-conda-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.6 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS/LAPACK: /home/llenezet/miniconda3/envs/env_pedixplorer/lib/libopenblasp-r0.3.27.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Europe/Paris</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] Pedixplorer_1.1.2 BiocStyle_2.32.1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] gtable_0.3.5        xfun_0.47           bslib_0.8.0        </span></span>
<span><span class="co">##  [4] ggplot2_3.5.1       htmlwidgets_1.6.4   lattice_0.22-6     </span></span>
<span><span class="co">##  [7] quadprog_1.5-8      vctrs_0.6.5         tools_4.4.1        </span></span>
<span><span class="co">## [10] generics_0.1.3      stats4_4.4.1        tibble_3.2.1       </span></span>
<span><span class="co">## [13] fansi_1.0.6         highr_0.11          pkgconfig_2.0.3    </span></span>
<span><span class="co">## [16] Matrix_1.6-5        data.table_1.15.4   desc_1.4.3         </span></span>
<span><span class="co">## [19] S4Vectors_0.40.2    readxl_1.4.3        lifecycle_1.0.4    </span></span>
<span><span class="co">## [22] stringr_1.5.1       compiler_4.4.1      shinytoastr_2.2.0  </span></span>
<span><span class="co">## [25] textshaping_0.4.0   munsell_0.5.1       httpuv_1.6.15      </span></span>
<span><span class="co">## [28] shinyWidgets_0.8.7  htmltools_0.5.8.1   sass_0.4.9         </span></span>
<span><span class="co">## [31] yaml_2.3.10         lazyeval_0.2.2      plotly_4.10.4      </span></span>
<span><span class="co">## [34] later_1.3.2         pillar_1.9.0        pkgdown_2.1.1      </span></span>
<span><span class="co">## [37] jquerylib_0.1.4     tidyr_1.3.1         DT_0.33            </span></span>
<span><span class="co">## [40] cachem_1.1.0        mime_0.12           tidyselect_1.2.1   </span></span>
<span><span class="co">## [43] digest_0.6.37       stringi_1.8.4       colourpicker_1.3.0 </span></span>
<span><span class="co">## [46] dplyr_1.1.4         purrr_1.0.2         bookdown_0.40      </span></span>
<span><span class="co">## [49] fastmap_1.2.0       grid_4.4.1          colorspace_2.1-1   </span></span>
<span><span class="co">## [52] cli_3.6.3           magrittr_2.0.3      utf8_1.2.4         </span></span>
<span><span class="co">## [55] withr_3.0.1         scales_1.3.0        promises_1.3.0     </span></span>
<span><span class="co">## [58] rmarkdown_2.28      httr_1.4.7          gridExtra_2.3      </span></span>
<span><span class="co">## [61] cellranger_1.1.0    ragg_1.3.3          shiny_1.9.1        </span></span>
<span><span class="co">## [64] evaluate_1.0.0      knitr_1.48          miniUI_0.1.1.1     </span></span>
<span><span class="co">## [67] viridisLite_0.4.2   rlang_1.1.4         Rcpp_1.0.13        </span></span>
<span><span class="co">## [70] xtable_1.8-4        glue_1.8.0          BiocManager_1.30.25</span></span>
<span><span class="co">## [73] BiocGenerics_0.48.1 jsonlite_1.8.9      R6_2.5.1           </span></span>
<span><span class="co">## [76] plyr_1.8.9          systemfonts_1.1.0   fs_1.6.4</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Louis Le Nézet, Jason Sinnwell, Terry Therneau.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
