<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Pedixplorer tutorial • Pedixplorer</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Pedixplorer tutorial">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Pedixplorer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/pedigree_alignment.html">Pedigree alignment details</a>
    </li>
    <li>
      <a href="../articles/pedigree_kinship.html">Pedigree kinship() details</a>
    </li>
    <li>
      <a href="../articles/pedigree_object.html">Pedigree object</a>
    </li>
    <li>
      <a href="../articles/pedigree_plot.html">Pedigree plotting details</a>
    </li>
    <li>
      <a href="../articles/pedigree.html">Pedixplorer tutorial</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/LouisLeNezet/Pedixplorer/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Pedixplorer tutorial</h1>
                        <h4 data-toc-skip class="author">Louis Le
Nézet</h4>
            
            <h4 data-toc-skip class="date">31/08/2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/LouisLeNezet/Pedixplorer/blob/change_workflow/vignettes/pedigree.Rmd" class="external-link"><code>vignettes/pedigree.Rmd</code></a></small>
      <div class="hidden name"><code>pedigree.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This document is a tutorial for the <code>Pedixplorer</code> package,
with examples of creating Pedigree objects and kinship matrices and
other Pedigree utilities.</p>
<p>The <code>Pedixplorer</code> package is an updated version of the <a href="https://github.com/mayoverse/kinship2" class="external-link"><code>Kinship2</code></a>
package, featuring a change in maintainer and repository from CRAN to
Bioconductor for continued development and support.</p>
<p>It contains the routines to handle family data with a Pedigree
object. The initial purpose was to create correlation structures that
describe family relationships such as kinship and identity-by-descent,
which can be used to model family data in mixed effects models, such as
in the coxme function. It also includes tools for Pedigree drawing and
filtering which is focused on producing compact layouts without
intervention. Recent additions include utilities to trim the Pedigree
object with various criteria, and kinship for the X chromosome.</p>
<p>Supplementary vignettes are available to explain:</p>
<ul>
<li>The
<strong><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mi>e</mi><mi>d</mi><mi>i</mi><mi>g</mi><mi>r</mi><mi>e</mi><mi>e</mi></mrow><annotation encoding="application/x-tex">Pedigree</annotation></semantics></math>
object</strong>
<code><a href="../articles/pedigree_object.html">vignette("pedigree_object", package = "Pedixplorer")</a></code>
</li>
<li>The <strong>alignment algorithm</strong> used create the Pedigree
structure
<code><a href="../articles/pedigree_alignment.html">vignette("pedigree_alignment", package = "Pedixplorer")</a></code>
</li>
<li>The <strong>kinship algorithm</strong>
<code><a href="../articles/pedigree_kinship.html">vignette("pedigree_kinship", package = "Pedixplorer")</a></code>
</li>
<li>The <strong>plotting algorithm</strong> used to plot the Pedigree
<code><a href="../articles/pedigree_plot.html">vignette("pedigree_plot", package = "Pedixplorer")</a></code>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mi>e</mi><mi>d</mi><mi>i</mi><mi>x</mi><mi>p</mi><mi>l</mi><mi>o</mi><mi>r</mi><mi>e</mi><mi>r</mi></mrow><annotation encoding="application/x-tex">Pedixplorer</annotation></semantics></math>
package is available on Bioconductor and can be installed with the
following command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"Pedixplorer"</span><span class="op">)</span></span></code></pre></div>
<p>The package can then be loaded with the following command:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LouisLeNezet/Pedixplorer" class="external-link">Pedixplorer</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="the-pedigree-s4-object">The Pedigree S4 object<a class="anchor" aria-label="anchor" href="#the-pedigree-s4-object"></a>
</h2>
<p>The
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mi>e</mi><mi>d</mi><mi>i</mi><mi>g</mi><mi>r</mi><mi>e</mi><mi>e</mi></mrow><annotation encoding="application/x-tex">Pedigree</annotation></semantics></math>
object is a list of dataframes that describe the family structure. It
contains the following components:</p>
<ul>
<li>ped: a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mi>e</mi><mi>d</mi></mrow><annotation encoding="application/x-tex">Ped</annotation></semantics></math>
object with the Pedigree information <code><a href="../reference/Ped-class.html">help(Ped)</a></code>.</li>
<li>rel: a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>R</mi><mi>e</mi><mi>l</mi></mrow><annotation encoding="application/x-tex">Rel</annotation></semantics></math>
object with the relationship information <code><a href="../reference/Rel-class.html">help(Rel)</a></code>.</li>
<li>scales: a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>S</mi><mi>c</mi><mi>a</mi><mi>l</mi><mi>e</mi><mi>s</mi></mrow><annotation encoding="application/x-tex">Scales</annotation></semantics></math>
object of 2 dataframe with the filling and borders informations for the
plot <code><a href="../reference/Scales-class.html">help(Scales)</a></code>.</li>
<li>hints: a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>H</mi><mi>i</mi><mi>n</mi><mi>t</mi><mi>s</mi></mrow><annotation encoding="application/x-tex">Hints</annotation></semantics></math>
objects with 2 slots indicating the horder and the spouse to organise
the Pedigree structure <code><a href="../reference/Hints-class.html">help(Hints)</a></code>.</li>
</ul>
</div>
<div class="section level2">
<h2 id="basic-usage">Basic Usage<a class="anchor" aria-label="anchor" href="#basic-usage"></a>
</h2>
<div class="section level3">
<h3 id="example-data">Example Data<a class="anchor" aria-label="anchor" href="#example-data"></a>
</h3>
<p>Two datasets are provided within the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mi>e</mi><mi>d</mi><mi>i</mi><mi>x</mi><mi>p</mi><mi>l</mi><mi>o</mi><mi>r</mi><mi>e</mi><mi>r</mi></mrow><annotation encoding="application/x-tex">Pedixplorer</annotation></semantics></math>
package: + minnbreast: 17 families from a breast cancer study +
sampleped: two sample pedigrees, with 41 and 14 subjects</p>
<p>This vignette uses the two pedigrees in
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>s</mi><mi>a</mi><mi>m</mi><mi>p</mi><mi>l</mi><mi>e</mi><mi>p</mi><mi>e</mi><mi>d</mi></mrow><annotation encoding="application/x-tex">sampleped</annotation></semantics></math>.
For more information on these datasets, see
<code><a href="../reference/minnbreast.html">help(minnbreast)</a></code> and <code><a href="../reference/sampleped.html">help(sampleped)</a></code>.</p>
</div>
<div class="section level3">
<h3 id="pedigree">Pedigree<a class="anchor" aria-label="anchor" href="#pedigree"></a>
</h3>
<p>First, we load
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>s</mi><mi>a</mi><mi>m</mi><mi>p</mi><mi>l</mi><mi>e</mi><mi>p</mi><mi>e</mi><mi>d</mi></mrow><annotation encoding="application/x-tex">sampleped</annotation></semantics></math>
and look at some of the values in the dataset, and create a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mi>e</mi><mi>d</mi><mi>i</mi><mi>g</mi><mi>r</mi><mi>e</mi><mi>e</mi></mrow><annotation encoding="application/x-tex">Pedigree</annotation></semantics></math>
object using the <code><a href="../reference/Pedigree-class.html">Pedigree()</a></code> function. This function
automaticaly detect the necessary columns in the dataframe. If necessary
you can modify the columns names with <em>cols_ren</em>. To create a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mi>e</mi><mi>d</mi><mi>i</mi><mi>g</mi><mi>r</mi><mi>e</mi><mi>e</mi></mrow><annotation encoding="application/x-tex">Pedigree</annotation></semantics></math>
object, with multiple families, the dataframe just need a family column
in the <em>ped_df</em> dataframe. When this is the case, the famid
column will be pasted to the id of each individuals separated by an
underscore to create a unique id for each individual in the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mi>e</mi><mi>d</mi><mi>i</mi><mi>g</mi><mi>r</mi><mi>e</mi><mi>e</mi></mrow><annotation encoding="application/x-tex">Pedigree</annotation></semantics></math>
object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"sampleped"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">sampleped</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    famid  id dadid momid sex affection avail num</span></span>
<span><span class="co">## 1      1 101  &lt;NA&gt;  &lt;NA&gt;   1         0     0   2</span></span>
<span><span class="co">## 2      1 102  &lt;NA&gt;  &lt;NA&gt;   2         1     0   3</span></span>
<span><span class="co">## 3      1 103   135   136   1         1     0   2</span></span>
<span><span class="co">## 4      1 104  &lt;NA&gt;  &lt;NA&gt;   2         0     0   4</span></span>
<span><span class="co">## 5      1 105  &lt;NA&gt;  &lt;NA&gt;   1        NA     0   6</span></span>
<span><span class="co">## 6      1 106  &lt;NA&gt;  &lt;NA&gt;   2        NA     0   1</span></span>
<span><span class="co">## 7      1 107  &lt;NA&gt;  &lt;NA&gt;   1         1     0  NA</span></span>
<span><span class="co">## 8      1 108  &lt;NA&gt;  &lt;NA&gt;   2         0     0   0</span></span>
<span><span class="co">## 9      1 109   101   102   2         0     1   3</span></span>
<span><span class="co">## 10     1 110   103   104   1         1     1   2</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ped</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Pedigree-class.html">Pedigree</a></span><span class="op">(</span><span class="va">sampleped</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">10</span>, <span class="fl">35</span>, <span class="fl">36</span><span class="op">)</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">ped</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Pedigree object with: </span></span>
<span><span class="co">## Ped object with 5 individuals and 13 metadata columns:</span></span>
<span><span class="co">##                    id       dadid       momid       sex       famid    steril    status     avail</span></span>
<span><span class="co">## col_class &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;ordered&gt; &lt;character&gt; &lt;logical&gt; &lt;logical&gt; &lt;logical&gt;</span></span>
<span><span class="co">## 1_103           1_103       1_135       1_136      male           1      &lt;NA&gt;      &lt;NA&gt;     FALSE</span></span>
<span><span class="co">## 1_104           1_104        &lt;NA&gt;        &lt;NA&gt;    female           1      &lt;NA&gt;      &lt;NA&gt;     FALSE</span></span>
<span><span class="co">## 1_110           1_110       1_103       1_104      male           1      &lt;NA&gt;      &lt;NA&gt;      TRUE</span></span>
<span><span class="co">## 1_135           1_135        &lt;NA&gt;        &lt;NA&gt;      male           1      &lt;NA&gt;      &lt;NA&gt;     FALSE</span></span>
<span><span class="co">## 1_136           1_136        &lt;NA&gt;        &lt;NA&gt;    female           1      &lt;NA&gt;      &lt;NA&gt;     FALSE</span></span>
<span><span class="co">##            affected    useful       kin     isinf num_child_tot num_child_dir num_child_ind |</span></span>
<span><span class="co">## col_class &lt;logical&gt; &lt;logical&gt; &lt;numeric&gt; &lt;logical&gt;     &lt;numeric&gt;     &lt;numeric&gt;     &lt;numeric&gt;  </span></span>
<span><span class="co">## 1_103          TRUE      &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;             1             1             0  </span></span>
<span><span class="co">## 1_104         FALSE      &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;             1             1             0  </span></span>
<span><span class="co">## 1_110          TRUE      &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;             0             0             0  </span></span>
<span><span class="co">## 1_135          &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;             1             1             0  </span></span>
<span><span class="co">## 1_136          &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;      &lt;NA&gt;             1             1             0  </span></span>
<span><span class="co">##                family       indId    fatherId    motherId      gender   affection   available</span></span>
<span><span class="co">## col_class &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">## 1_103               1         103         135         136           1           1           0</span></span>
<span><span class="co">## 1_104               1         104        &lt;NA&gt;        &lt;NA&gt;           2           0           0</span></span>
<span><span class="co">## 1_110               1         110         103         104           1           1           1</span></span>
<span><span class="co">## 1_135               1         135        &lt;NA&gt;        &lt;NA&gt;           1        &lt;NA&gt;           0</span></span>
<span><span class="co">## 1_136               1         136        &lt;NA&gt;        &lt;NA&gt;           2        &lt;NA&gt;           0</span></span>
<span><span class="co">##                   num       error sterilisation vitalStatus affection_mods  avail_mods</span></span>
<span><span class="co">## col_class &lt;character&gt; &lt;character&gt;   &lt;character&gt; &lt;character&gt;    &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">## 1_103               2        &lt;NA&gt;          &lt;NA&gt;        &lt;NA&gt;              1           0</span></span>
<span><span class="co">## 1_104               4        &lt;NA&gt;          &lt;NA&gt;        &lt;NA&gt;              0           0</span></span>
<span><span class="co">## 1_110               2        &lt;NA&gt;          &lt;NA&gt;        &lt;NA&gt;              1           1</span></span>
<span><span class="co">## 1_135               5        &lt;NA&gt;          &lt;NA&gt;        &lt;NA&gt;             NA           0</span></span>
<span><span class="co">## 1_136               6        &lt;NA&gt;          &lt;NA&gt;        &lt;NA&gt;             NA           0</span></span>
<span><span class="co">## Rel object with 0 relationshipswith 0 MZ twin, 0 DZ twin, 0 UZ twin, 0 Spouse:</span></span>
<span><span class="co">##          id1         id2                     code       famid</span></span>
<span><span class="co">##  &lt;character&gt; &lt;character&gt; &lt;c("ordered", "factor")&gt; &lt;character&gt;</span></span></code></pre>
<p>For more information on the <code><a href="../reference/Pedigree-class.html">Pedigree()</a></code> function, see
<code><a href="../reference/Pedigree-class.html">help(Pedigree)</a></code>.</p>
<p>The
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mi>e</mi><mi>d</mi><mi>i</mi><mi>g</mi><mi>r</mi><mi>e</mi><mi>e</mi></mrow><annotation encoding="application/x-tex">Pedigree</annotation></semantics></math>
object can be subset to individual pedigrees by their family id. The
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mi>e</mi><mi>d</mi><mi>i</mi><mi>g</mi><mi>r</mi><mi>e</mi><mi>e</mi></mrow><annotation encoding="application/x-tex">Pedigree</annotation></semantics></math>
object has a print, summary and plot method, which we show below. The
print method prints the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mi>e</mi><mi>d</mi></mrow><annotation encoding="application/x-tex">Ped</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>R</mi><mi>e</mi><mi>l</mi></mrow><annotation encoding="application/x-tex">Rel</annotation></semantics></math>
object of the pedigree. The summary method prints a short summary of the
pedigree. Finally the plot method displays the pedigree.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ped</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Pedigree-class.html">Pedigree</a></span><span class="op">(</span><span class="va">sampleped</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="../reference/Ped-class.html">famid</a></span><span class="op">(</span><span class="fu"><a href="../reference/Pedigree-class.html">ped</a></span><span class="op">(</span><span class="va">ped</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "1" "1" "1" "1" "1" "1" "1" "1" "1" "1" "1" "1" "1" "1" "1" "1" "1" "1" "1" "1" "1" "1" "1" "1"</span></span>
<span><span class="co">## [25] "1" "1" "1" "1" "1" "1" "1" "1" "1" "1" "1" "1" "1" "1" "1" "1" "1" "2" "2" "2" "2" "2" "2" "2"</span></span>
<span><span class="co">## [49] "2" "2" "2" "2" "2" "2" "2"</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ped1</span> <span class="op">&lt;-</span> <span class="va">ped</span><span class="op">[</span><span class="fu"><a href="../reference/Ped-class.html">famid</a></span><span class="op">(</span><span class="fu"><a href="../reference/Pedigree-class.html">ped</a></span><span class="op">(</span><span class="va">ped</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="st">"1"</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">ped1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Pedigree object with </span></span>
<span><span class="co">## [1] "Ped object with 41 individuals and 13 metadata columns"</span></span>
<span><span class="co">## [1] "Rel object with 0 relationshipswith 0 MZ twin, 0 DZ twin, 0 UZ twin, 0 Spouse"</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">ped1</span><span class="op">)</span></span></code></pre></div>
<p><img src="pedigree_files/figure-html/ped1-1.png" alt="Pedigree of family 1" width="700"></p>
<p>You can add a title and a legend to the plot with the following
command:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">ped1</span>, title <span class="op">=</span> <span class="st">"Pedigree 1"</span>, legend <span class="op">=</span> <span class="cn">TRUE</span>, leg_loc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">15</span>, <span class="fl">4.5</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="pedigree_files/figure-html/ped1_title-1.png" alt="Pedigree of family 1 with legend" width="700">
# Pedigree Shiny application</p>
<p>A shiny application is available to create, interact and plot
pedigrees. To launch the application, use the following command:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/interactive.html" class="external-link">interactive</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/ped_shiny.html">ped_shiny</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>The use is simple:</p>
<ul>
<li>You first need to import a dataset and select the columns to
use.</li>
<li>You can then select the affection informations and the colors
associated to them.</li>
<li>If different families are present in the dataset, you can select
which one to plot.</li>
<li>Before the plot is displayed, you can filter the pedigree by
selecting the informatives subjects to keep and their relatives. If the
pedigree is then splited in multiple families, you can select which to
plot.</li>
<li>Finally the plot is displayed and you can make it interactive and
download the resulting image.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="fixing-pedigree-issues">Fixing Pedigree Issues<a class="anchor" aria-label="anchor" href="#fixing-pedigree-issues"></a>
</h2>
<p>To “break” the pedigree, we can manipulate the sex value to not match
the parent value (in this example, we change
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>203</mn><annotation encoding="application/x-tex">203</annotation></semantics></math>
from a male to a female, even though
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>203</mn><annotation encoding="application/x-tex">203</annotation></semantics></math>
is a father). To do this, we first subset
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>d</mi><mi>a</mi><mi>t</mi><mi>p</mi><mi>e</mi><mi>d</mi><mn>2</mn></mrow><annotation encoding="application/x-tex">datped2</annotation></semantics></math>,
locate the <em>id</em> column, and match it to a specific id (in this
case,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>203</mn><annotation encoding="application/x-tex">203</annotation></semantics></math>).
Within id
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>203</mn><annotation encoding="application/x-tex">203</annotation></semantics></math>,
then locate in the <em>sex</em> column. Assign this subset to the
incorrect value of <em>2</em> (female) to change the original/correct
value of <em>1</em> (male).</p>
<p>To further break the pedigree, we can delete subjects who seem
irrelevant to the pedigree (in this example, we delete
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>209</mn><annotation encoding="application/x-tex">209</annotation></semantics></math>
because he is a married-in father). To do this, we subset
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>d</mi><mi>a</mi><mi>t</mi><mi>p</mi><mi>e</mi><mi>d</mi><mn>2</mn></mrow><annotation encoding="application/x-tex">datped2</annotation></semantics></math>
and use the <em>-which()</em> function to locate and delete the
specified subject (in this case,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>209</mn><annotation encoding="application/x-tex">209</annotation></semantics></math>).
Reassign this code to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>d</mi><mi>a</mi><mi>t</mi><mi>p</mi><mi>e</mi><mi>d</mi><mn>22</mn></mrow><annotation encoding="application/x-tex">datped22</annotation></semantics></math>
to drop the specified subject entirely.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">datped2</span> <span class="op">&lt;-</span> <span class="va">sampleped</span><span class="op">[</span><span class="va">sampleped</span><span class="op">$</span><span class="va">famid</span> <span class="op">==</span> <span class="fl">2</span>, <span class="op">]</span></span>
<span><span class="va">datped2</span><span class="op">[</span><span class="va">datped2</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fl">203</span>, <span class="st">"sex"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span><span class="va">datped2</span> <span class="op">&lt;-</span> <span class="va">datped2</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">datped2</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fl">209</span><span class="op">)</span>, <span class="op">]</span></span></code></pre></div>
<p>An error occurs when the <code><a href="../reference/Pedigree-class.html">Pedigree()</a></code> function notices
that id
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>203</mn><annotation encoding="application/x-tex">203</annotation></semantics></math>
is not coded to be male (<em>1</em>) but is a father. To correct this,
we simply employ the <code><a href="../reference/fix_parents.html">fix_parents()</a></code> function to adjust the
<em>sex</em> value to match either <em>momid</em> or <em>dadid</em>.
<code><a href="../reference/fix_parents.html">fix_parents()</a></code> will also add back in any deleted subjects,
further fixing the Pedigree.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tryout</span> <span class="op">&lt;-</span> <span class="kw"><a href="https://rdrr.io/r/base/try.html" class="external-link">try</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="va">ped2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Pedigree-class.html">Pedigree</a></span><span class="op">(</span><span class="va">datped2</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Error in validObject(.Object) : </span></span>
<span><span class="co">##   invalid class "Ped" object: dadid values '2_209' should be in '2_201', '2_202', '2_203', '2_204', '2_205'...</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fixped2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">datped2</span>, <span class="fu"><a href="../reference/fix_parents.html">fix_parents</a></span><span class="op">(</span><span class="va">id</span>, <span class="va">dadid</span>, <span class="va">momid</span>, <span class="va">sex</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fixped2</span></span></code></pre></div>
<pre><code><span><span class="co">##     id momid dadid sex famid</span></span>
<span><span class="co">## 1  201  &lt;NA&gt;  &lt;NA&gt;   1     1</span></span>
<span><span class="co">## 2  202  &lt;NA&gt;  &lt;NA&gt;   2     1</span></span>
<span><span class="co">## 3  203  &lt;NA&gt;  &lt;NA&gt;   1     1</span></span>
<span><span class="co">## 4  204   202   201   2     1</span></span>
<span><span class="co">## 5  205   202   201   1     1</span></span>
<span><span class="co">## 6  206   202   201   2     1</span></span>
<span><span class="co">## 7  207   202   201   2     1</span></span>
<span><span class="co">## 8  208   202   201   2     1</span></span>
<span><span class="co">## 9  210   204   203   1     1</span></span>
<span><span class="co">## 10 211   204   203   1     1</span></span>
<span><span class="co">## 11 212   208   209   2     1</span></span>
<span><span class="co">## 12 213   208   209   1     1</span></span>
<span><span class="co">## 13 214   208   209   1     1</span></span>
<span><span class="co">## 14 209  &lt;NA&gt;  &lt;NA&gt;   1     1</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ped2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Pedigree-class.html">Pedigree</a></span><span class="op">(</span><span class="va">fixped2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">ped2</span><span class="op">)</span></span></code></pre></div>
<p><img src="pedigree_files/figure-html/fixped2-1.png" alt="Pedigree of family 2" width="700"></p>
<p>If the fix is straightforward (changing one sex value based on either
being a mother or father), <code><a href="../reference/fix_parents.html">fix_parents()</a></code> will resolve the
issue. If the issue is more complicated, say if
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>203</mn><annotation encoding="application/x-tex">203</annotation></semantics></math>
is coded to be both a father and a mother, <code><a href="../reference/fix_parents.html">fix_parents()</a></code>
will not know which one is correct and therefore the issue will not be
resolved.</p>
</div>
<div class="section level2">
<h2 id="kinship">Kinship<a class="anchor" aria-label="anchor" href="#kinship"></a>
</h2>
<p>A common use for pedigrees is to make a matrix of kinship
coefficients that can be used in mixed effect models. A kinship
coefficient is the probability that a randomly selected allele from two
people at a given locus will be identical by descent (IBD), assuming all
founder alleles are independent. For example, we each have two alleles
per autosomal marker, so sampling two alleles with replacement from our
own DNA has only
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mo>=</mo><mn>0.50</mn></mrow><annotation encoding="application/x-tex">p=0.50</annotation></semantics></math>
probability of getting the same allele twice.</p>
<div class="section level3">
<h3 id="kinship-for-pedigree-object">Kinship for Pedigree object<a class="anchor" aria-label="anchor" href="#kinship-for-pedigree-object"></a>
</h3>
<p>We use <code><a href="../reference/kinship.html">kinship()</a></code> to calculate the kinship matrix for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mi>e</mi><mi>d</mi><mn>2</mn></mrow><annotation encoding="application/x-tex">ped2</annotation></semantics></math>.
The result is a special symmetrix matrix class from the <a href="https://CRAN.R-project.org/package=Matrix/" class="external-link">Matrix R package</a>,
which is stored efficiently to avoid repeating elements.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kin2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/kinship.html">kinship</a></span><span class="op">(</span><span class="va">ped2</span><span class="op">)</span></span>
<span><span class="va">kin2</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## 9 x 9 sparse Matrix of class "dsCMatrix"</span></span>
<span><span class="co">##       1_201 1_202 1_203 1_204 1_205 1_206 1_207 1_208 1_209</span></span>
<span><span class="co">## 1_201  0.50  .      .    0.25  0.25  0.25  0.25  0.25   .  </span></span>
<span><span class="co">## 1_202  .     0.50   .    0.25  0.25  0.25  0.25  0.25   .  </span></span>
<span><span class="co">## 1_203  .     .      0.5  .     .     .     .     .      .  </span></span>
<span><span class="co">## 1_204  0.25  0.25   .    0.50  0.25  0.25  0.25  0.25   .  </span></span>
<span><span class="co">## 1_205  0.25  0.25   .    0.25  0.50  0.25  0.25  0.25   .  </span></span>
<span><span class="co">## 1_206  0.25  0.25   .    0.25  0.25  0.50  0.25  0.25   .  </span></span>
<span><span class="co">## 1_207  0.25  0.25   .    0.25  0.25  0.25  0.50  0.25   .  </span></span>
<span><span class="co">## 1_208  0.25  0.25   .    0.25  0.25  0.25  0.25  0.50   .  </span></span>
<span><span class="co">## 1_209  .     .      .    .     .     .     .     .      0.5</span></span></code></pre>
<p>For family 2, see that the row and column names match the id in the
figure below, and see that each kinship coefficient with themselves is
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>0.50</mn><annotation encoding="application/x-tex">0.50</annotation></semantics></math>,
siblings are
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>0.25</mn><annotation encoding="application/x-tex">0.25</annotation></semantics></math>
(e.g. <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>204</mn><mo>−</mo><mn>205</mn></mrow><annotation encoding="application/x-tex">204-205</annotation></semantics></math>),
and pedigree marry-ins only share alleles IBD with their children with
coefficient
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>0.25</mn><annotation encoding="application/x-tex">0.25</annotation></semantics></math>
(e.g.
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>203</mn><mo>−</mo><mn>210</mn></mrow><annotation encoding="application/x-tex">203-210</annotation></semantics></math>).
The plot can be used to verify other kinship coefficients.</p>
</div>
<div class="section level3">
<h3 id="kinship-for-pedigree-with-multiple-families">Kinship for Pedigree with multiple families<a class="anchor" aria-label="anchor" href="#kinship-for-pedigree-with-multiple-families"></a>
</h3>
<p>The <code><a href="../reference/kinship.html">kinship()</a></code> function also works on a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mi>e</mi><mi>d</mi><mi>i</mi><mi>g</mi><mi>r</mi><mi>e</mi><mi>e</mi></mrow><annotation encoding="application/x-tex">Pedigree</annotation></semantics></math>
object with multiple families. We show how to create the kinship matrix,
then show a snapshot of them for the two families, where the row and
columns names are the ids of the subject.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ped</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Pedigree-class.html">Pedigree</a></span><span class="op">(</span><span class="va">sampleped</span><span class="op">)</span></span>
<span><span class="va">kin_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/kinship.html">kinship</a></span><span class="op">(</span><span class="va">ped</span><span class="op">)</span></span>
<span><span class="va">kin_all</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## 9 x 9 sparse Matrix of class "dsCMatrix"</span></span>
<span><span class="co">##       1_101 1_102 1_103 1_104 1_105 1_106 1_107 1_108 1_109</span></span>
<span><span class="co">## 1_101  0.50  .      .     .     .     .     .     .    0.25</span></span>
<span><span class="co">## 1_102  .     0.50   .     .     .     .     .     .    0.25</span></span>
<span><span class="co">## 1_103  .     .      0.5   .     .     .     .     .    .   </span></span>
<span><span class="co">## 1_104  .     .      .     0.5   .     .     .     .    .   </span></span>
<span><span class="co">## 1_105  .     .      .     .     0.5   .     .     .    .   </span></span>
<span><span class="co">## 1_106  .     .      .     .     .     0.5   .     .    .   </span></span>
<span><span class="co">## 1_107  .     .      .     .     .     .     0.5   .    .   </span></span>
<span><span class="co">## 1_108  .     .      .     .     .     .     .     0.5  .   </span></span>
<span><span class="co">## 1_109  0.25  0.25   .     .     .     .     .     .    0.50</span></span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kin_all</span><span class="op">[</span><span class="fl">40</span><span class="op">:</span><span class="fl">43</span>, <span class="fl">40</span><span class="op">:</span><span class="fl">43</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## 4 x 4 sparse Matrix of class "dsCMatrix"</span></span>
<span><span class="co">##       1_140 1_141 2_201 2_202</span></span>
<span><span class="co">## 1_140  0.50  0.25   .     .  </span></span>
<span><span class="co">## 1_141  0.25  0.50   .     .  </span></span>
<span><span class="co">## 2_201  .     .      0.5   .  </span></span>
<span><span class="co">## 2_202  .     .      .     0.5</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kin_all</span><span class="op">[</span><span class="fl">42</span><span class="op">:</span><span class="fl">46</span>, <span class="fl">42</span><span class="op">:</span><span class="fl">46</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## 5 x 5 sparse Matrix of class "dsCMatrix"</span></span>
<span><span class="co">##       2_201 2_202 2_203 2_204 2_205</span></span>
<span><span class="co">## 2_201  0.50  .      .    0.25  0.25</span></span>
<span><span class="co">## 2_202  .     0.50   .    0.25  0.25</span></span>
<span><span class="co">## 2_203  .     .      0.5  .     .   </span></span>
<span><span class="co">## 2_204  0.25  0.25   .    0.50  0.25</span></span>
<span><span class="co">## 2_205  0.25  0.25   .    0.25  0.50</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="kinship-for-twins-in-pedigree-with-multiple-families">Kinship for twins in Pedigree with multiple families<a class="anchor" aria-label="anchor" href="#kinship-for-twins-in-pedigree-with-multiple-families"></a>
</h3>
<p>Specifying twin relationships in a Pedigree with multiple families
object is complicated by the fact that the user must specify the family
id to which the <em>id1</em> and <em>id2</em> belong. We show below the
relation matrix requires the family id to be in the last column, with
the column names as done below, to make the plotting and kinship
matrices to show up with the monozygotic twins correctly. We show how to
specify monozygosity for subjects
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>206</mn><annotation encoding="application/x-tex">206</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>207</mn><annotation encoding="application/x-tex">207</annotation></semantics></math>
in
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mi>e</mi><mi>d</mi><mn>2</mn></mrow><annotation encoding="application/x-tex">ped2</annotation></semantics></math>,
and subjects
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>125</mn><annotation encoding="application/x-tex">125</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>126</mn><annotation encoding="application/x-tex">126</annotation></semantics></math>
in
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mi>e</mi><mi>d</mi><mn>1</mn></mrow><annotation encoding="application/x-tex">ped1</annotation></semantics></math>.
We check it by looking at the kinship matrix for these pairs, which are
correctly at
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>0.5</mn><annotation encoding="application/x-tex">0.5</annotation></semantics></math>.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"relped"</span><span class="op">)</span></span>
<span><span class="va">relped</span></span></code></pre></div>
<pre><code><span><span class="co">##   famid id1 id2 code</span></span>
<span><span class="co">## 1     1 140 141    1</span></span>
<span><span class="co">## 2     1 139 140    2</span></span>
<span><span class="co">## 3     1 121 123    2</span></span>
<span><span class="co">## 4     1 129 126    4</span></span>
<span><span class="co">## 5     1 130 133    3</span></span>
<span><span class="co">## 6     2 210 211    1</span></span>
<span><span class="co">## 7     2 208 204    2</span></span>
<span><span class="co">## 8     2 212 213    3</span></span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ped</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Pedigree-class.html">Pedigree</a></span><span class="op">(</span><span class="va">sampleped</span>, <span class="va">relped</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kin_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/kinship.html">kinship</a></span><span class="op">(</span><span class="va">ped</span><span class="op">)</span></span>
<span><span class="va">kin_all</span><span class="op">[</span><span class="fl">24</span><span class="op">:</span><span class="fl">27</span>, <span class="fl">24</span><span class="op">:</span><span class="fl">27</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## 4 x 4 sparse Matrix of class "dsCMatrix"</span></span>
<span><span class="co">##        1_124  1_125  1_126  1_127</span></span>
<span><span class="co">## 1_124 0.5000 0.0625 0.0625 0.0625</span></span>
<span><span class="co">## 1_125 0.0625 0.5000 0.2500 0.1250</span></span>
<span><span class="co">## 1_126 0.0625 0.2500 0.5000 0.1250</span></span>
<span><span class="co">## 1_127 0.0625 0.1250 0.1250 0.5000</span></span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kin_all</span><span class="op">[</span><span class="fl">46</span><span class="op">:</span><span class="fl">50</span>, <span class="fl">46</span><span class="op">:</span><span class="fl">50</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## 5 x 5 sparse Matrix of class "dsCMatrix"</span></span>
<span><span class="co">##       2_205 2_206 2_207 2_208 2_209</span></span>
<span><span class="co">## 2_205  0.50  0.25  0.25  0.25   .  </span></span>
<span><span class="co">## 2_206  0.25  0.50  0.25  0.25   .  </span></span>
<span><span class="co">## 2_207  0.25  0.25  0.50  0.25   .  </span></span>
<span><span class="co">## 2_208  0.25  0.25  0.25  0.50   .  </span></span>
<span><span class="co">## 2_209  .     .     .     .      0.5</span></span></code></pre>
<p>Note that subject
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>113</mn><annotation encoding="application/x-tex">113</annotation></semantics></math>
is not in
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mi>e</mi><mi>d</mi><mn>1</mn></mrow><annotation encoding="application/x-tex">ped1</annotation></semantics></math>
because they are a marry-in without children in the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mi>e</mi><mi>d</mi><mi>i</mi><mi>g</mi><mi>r</mi><mi>e</mi><mi>e</mi></mrow><annotation encoding="application/x-tex">Pedigree</annotation></semantics></math>.
Subject
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>113</mn><annotation encoding="application/x-tex">113</annotation></semantics></math>
is in their own
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mi>e</mi><mi>d</mi><mi>i</mi><mi>g</mi><mi>r</mi><mi>e</mi><mi>e</mi></mrow><annotation encoding="application/x-tex">Pedigree</annotation></semantics></math>
of size 1 in the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>k</mi><mi>i</mi><msub><mi>n</mi><mi>a</mi></msub><mi>l</mi><mi>l</mi></mrow><annotation encoding="application/x-tex">kin_all</annotation></semantics></math>
matrix at index
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>41</mn><annotation encoding="application/x-tex">41</annotation></semantics></math>.
We later show how to handle such marry-ins for plotting.</p>
</div>
</div>
<div class="section level2">
<h2 id="optional-pedigree-informations">Optional Pedigree Informations<a class="anchor" aria-label="anchor" href="#optional-pedigree-informations"></a>
</h2>
<p>We use
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mi>e</mi><mi>d</mi><mn>2</mn></mrow><annotation encoding="application/x-tex">ped2</annotation></semantics></math>
from
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>s</mi><mi>a</mi><mi>m</mi><mi>p</mi><mi>l</mi><mi>e</mi><mi>p</mi><mi>e</mi><mi>d</mi></mrow><annotation encoding="application/x-tex">sampleped</annotation></semantics></math>
to sequentially add optional information to the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mi>e</mi><mi>d</mi><mi>i</mi><mi>g</mi><mi>r</mi><mi>e</mi><mi>e</mi></mrow><annotation encoding="application/x-tex">Pedigree</annotation></semantics></math>
object.</p>
<div class="section level3">
<h3 id="status">Status<a class="anchor" aria-label="anchor" href="#status"></a>
</h3>
<p>The example below shows how to specify a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>s</mi><mi>t</mi><mi>a</mi><mi>t</mi><mi>u</mi><mi>s</mi></mrow><annotation encoding="application/x-tex">status</annotation></semantics></math>
indicator, such as vital status. The
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>s</mi><mi>a</mi><mi>m</mi><mi>p</mi><mi>l</mi><mi>e</mi><mi>p</mi><mi>e</mi><mi>d</mi></mrow><annotation encoding="application/x-tex">sampleped</annotation></semantics></math>
data does not include such an<br>
indicator, so we create one to indicate that the first generation of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mi>e</mi><mi>d</mi><mn>2</mn></mrow><annotation encoding="application/x-tex">ped2</annotation></semantics></math>,
subjects
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>1</mn><annotation encoding="application/x-tex">1</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>2</mn><annotation encoding="application/x-tex">2</annotation></semantics></math>,
are deceased. The
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>s</mi><mi>t</mi><mi>a</mi><mi>t</mi><mi>u</mi><mi>s</mi></mrow><annotation encoding="application/x-tex">status</annotation></semantics></math>
indicator is used to cross out the individuals in the Pedigree plot.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df2</span> <span class="op">&lt;-</span> <span class="va">sampleped</span><span class="op">[</span><span class="va">sampleped</span><span class="op">$</span><span class="va">famid</span> <span class="op">==</span> <span class="fl">2</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">df2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "famid"     "id"        "dadid"     "momid"     "sex"       "affection" "avail"     "num"</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df2</span><span class="op">$</span><span class="va">status</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">12</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ped2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Pedigree-class.html">Pedigree</a></span><span class="op">(</span><span class="va">df2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="../reference/Ped-class.html">status</a></span><span class="op">(</span><span class="fu"><a href="../reference/Pedigree-class.html">ped</a></span><span class="op">(</span><span class="va">ped2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    Mode   FALSE    TRUE </span></span>
<span><span class="co">## logical      12       2</span></span></code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">ped2</span><span class="op">)</span></span></code></pre></div>
<p><img src="pedigree_files/figure-html/status-1.png" alt="Pedigree of family 2 with different vital status" width="700"></p>
</div>
<div class="section level3">
<h3 id="labels">Labels<a class="anchor" aria-label="anchor" href="#labels"></a>
</h3>
<p>Here we show how to use the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>l</mi><mi>a</mi><mi>b</mi><mi>e</mi><mi>l</mi></mrow><annotation encoding="application/x-tex">label</annotation></semantics></math>
argument in the plot method to add additional information under each
subject. In the example below, we add names to the existing plot by
adding a new column to the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>e</mi><mi>l</mi><mi>e</mi><mi>m</mi><mi>e</mi><mi>n</mi><mi>t</mi><mi>M</mi><mi>e</mi><mi>t</mi><mi>a</mi><mi>d</mi><mi>a</mi><mi>t</mi><mi>a</mi></mrow><annotation encoding="application/x-tex">elementMetadata</annotation></semantics></math>
of the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mi>e</mi><mi>d</mi></mrow><annotation encoding="application/x-tex">Ped</annotation></semantics></math>
object of the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mi>e</mi><mi>d</mi><mi>i</mi><mi>g</mi><mi>r</mi><mi>e</mi><mi>e</mi></mrow><annotation encoding="application/x-tex">Pedigree</annotation></semantics></math>.</p>
<p>As space permits, more lines and characters per line can be made
using the a {/em } character to indicate a new line.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mcols</span><span class="op">(</span><span class="va">ped2</span><span class="op">)</span><span class="op">$</span><span class="va">Names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"John\nDalton"</span>, <span class="st">"Linda"</span>, <span class="st">"Jack"</span>, <span class="st">"Rachel"</span>, <span class="st">"Joe"</span>, <span class="st">"Deb"</span>,</span>
<span>    <span class="st">"Lucy"</span>, <span class="st">"Ken"</span>, <span class="st">"Barb"</span>, <span class="st">"Mike"</span>, <span class="st">"Matt"</span>,</span>
<span>    <span class="st">"Mindy"</span>, <span class="st">"Mark"</span>, <span class="st">"Marie\nCurie"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">ped2</span>, label <span class="op">=</span> <span class="st">"Names"</span><span class="op">)</span></span></code></pre></div>
<p><img src="pedigree_files/figure-html/labels-1.png" alt="Pedigree of family 2 with names label" width="700"></p>
</div>
<div class="section level3">
<h3 id="affected-indicators">Affected Indicators<a class="anchor" aria-label="anchor" href="#affected-indicators"></a>
</h3>
<p>We show how to specify affected status with a single indicator and
multiple indicators. First, we use the affected indicator from
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>s</mi><mi>a</mi><mi>m</mi><mi>p</mi><mi>l</mi><mi>e</mi><mi>p</mi><mi>e</mi><mi>d</mi></mrow><annotation encoding="application/x-tex">sampleped</annotation></semantics></math>,
which contains
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>0</mn><annotation encoding="application/x-tex">0</annotation></semantics></math>/<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>1</mn><annotation encoding="application/x-tex">1</annotation></semantics></math>
indicators and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>N</mi><mi>A</mi></mrow><annotation encoding="application/x-tex">NA</annotation></semantics></math>
as missing, and let it it indicate blue eyes. Next, we create a vector
as an indicator for baldness. And add it as a second filling scale for
the plot with <code>generate_colors(add_to_scale = TRUE)</code>. The
plot shapes for each subject is therefore divided into two equal parts
and shaded differently to indicate the two affected indicators.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mcols</span><span class="op">(</span><span class="va">ped2</span><span class="op">)</span><span class="op">$</span><span class="va">bald</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ped2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generate_colors.html">generate_colors</a></span><span class="op">(</span><span class="va">ped2</span>, col_aff <span class="op">=</span> <span class="st">"bald"</span>, add_to_scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">ped2</span>, legend <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="pedigree_files/figure-html/two_affection-1.png" alt="Pedigree of family 2 with two affection indicators" width="700"></p>
</div>
<div class="section level3">
<h3 id="special-relationships">Special Relationships<a class="anchor" aria-label="anchor" href="#special-relationships"></a>
</h3>
<p>Special pedigree relationships can be specified in a matrix as the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>r</mi><mi>e</mi><mi>l</mi><mi>a</mi><mi>t</mi><mi>i</mi><mi>o</mi><mi>n</mi></mrow><annotation encoding="application/x-tex">relation</annotation></semantics></math>
argument. There are 4 relationships that can be specified by numeric
codes:</p>
<ul>
<li>1 = Monozygotic twins</li>
<li>2 = Dizygotic twins</li>
<li>3 = twins of unknown zygosity</li>
<li>4 = Spouse</li>
</ul>
<p>The spouse relationship can indicate a marry-in when a couple does
not have children together.</p>
<div class="section level4">
<h4 id="twins">Twins<a class="anchor" aria-label="anchor" href="#twins"></a>
</h4>
<p>Below, we use the relationship dataset. We can specify in the code
column if the individuals are monozygotic ‘1’, dizygotic ‘2’ or of
unknown-zygosity ‘3’ twins. The twin relationships are both represented
with diverging lines from a single point. The monozygotic twins have an
additional line connecting the diverging lines, while twins of unknown
zygosity have a question mark.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## create twin relationships</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"relped"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Pedigree-class.html">rel</a></span><span class="op">(</span><span class="va">ped2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Rel-class.html">Rel</a></span><span class="op">(</span><span class="va">relped</span><span class="op">[</span><span class="va">relped</span><span class="op">$</span><span class="va">famid</span> <span class="op">==</span> <span class="fl">2</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">ped2</span><span class="op">)</span></span></code></pre></div>
<p><img src="pedigree_files/figure-html/twins-1.png" alt="Pedigree of family 2 with special relationships" width="700"></p>
</div>
<div class="section level4">
<h4 id="inbreeding">Inbreeding<a class="anchor" aria-label="anchor" href="#inbreeding"></a>
</h4>
<p>Another special relationship is inbreeding. Inbreeding of founders
implies the founders’ parents are related (the maternal and paternal
genes descended from a single ancestral gene). One thing we can do is
add more people to the pedigree to show this inbreeding.</p>
<p>To show that a pair of founders (subjects
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>201</mn><annotation encoding="application/x-tex">201</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>202</mn><annotation encoding="application/x-tex">202</annotation></semantics></math>)
are inbred, we must show that their parents are siblings. To do this, we
create subjects
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>197</mn><annotation encoding="application/x-tex">197</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>198</mn><annotation encoding="application/x-tex">198</annotation></semantics></math>
to be the parents of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>201</mn><annotation encoding="application/x-tex">201</annotation></semantics></math>
and also create subjects
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>199</mn><annotation encoding="application/x-tex">199</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>200</mn><annotation encoding="application/x-tex">200</annotation></semantics></math>
to be the parents of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>202</mn><annotation encoding="application/x-tex">202</annotation></semantics></math>.
To make subjects
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>198</mn><annotation encoding="application/x-tex">198</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>199</mn><annotation encoding="application/x-tex">199</annotation></semantics></math>
siblings, we give them the same parents, creating subjects
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>195</mn><annotation encoding="application/x-tex">195</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>196</mn><annotation encoding="application/x-tex">196</annotation></semantics></math>.
This results in subjects
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>201</mn><annotation encoding="application/x-tex">201</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>202</mn><annotation encoding="application/x-tex">202</annotation></semantics></math>
being first cousins, and therefore inbred.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">indid</span> <span class="op">&lt;-</span> <span class="fl">195</span><span class="op">:</span><span class="fl">202</span></span>
<span><span class="va">dadid</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="cn">NA</span>, <span class="cn">NA</span>, <span class="fl">196</span>, <span class="fl">196</span>, <span class="cn">NA</span>, <span class="fl">197</span>, <span class="fl">199</span><span class="op">)</span></span>
<span><span class="va">momid</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="cn">NA</span>, <span class="cn">NA</span>, <span class="fl">195</span>, <span class="fl">195</span>, <span class="cn">NA</span>, <span class="fl">198</span>, <span class="fl">200</span><span class="op">)</span></span>
<span><span class="va">sex</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">ped3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    id <span class="op">=</span> <span class="va">indid</span>, dadid <span class="op">=</span> <span class="va">dadid</span>,</span>
<span>    momid <span class="op">=</span> <span class="va">momid</span>, sex <span class="op">=</span> <span class="va">sex</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">ped4df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind.data.frame</a></span><span class="op">(</span><span class="va">df2</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, <span class="fl">2</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>, <span class="va">ped3</span><span class="op">)</span></span>
<span><span class="va">ped4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Pedigree-class.html">Pedigree</a></span><span class="op">(</span><span class="va">ped4df</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">ped4</span><span class="op">)</span></span></code></pre></div>
<p><img src="pedigree_files/figure-html/inbreeding-1.png" alt="Pedigree with inbreeding" width="700"></p>
</div>
<div class="section level4">
<h4 id="marry-ins">Marry-ins<a class="anchor" aria-label="anchor" href="#marry-ins"></a>
</h4>
<p>Spouse with no child can also be specified with the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>r</mi><mi>e</mi><msub><mi>l</mi><mi>d</mi></msub><mi>f</mi></mrow><annotation encoding="application/x-tex">rel_df</annotation></semantics></math>
argument by setting the code value to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>S</mi><mi>p</mi><mi>o</mi><mi>u</mi><mi>s</mi><mi>e</mi></mrow><annotation encoding="application/x-tex">Spouse</annotation></semantics></math>
or
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>4</mn><annotation encoding="application/x-tex">4</annotation></semantics></math>.
If we use the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mi>e</mi><mi>d</mi><mn>2</mn></mrow><annotation encoding="application/x-tex">ped2</annotation></semantics></math>
from earlier and add a new spouse relationship between the individuals
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>212</mn><annotation encoding="application/x-tex">212</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>211</mn><annotation encoding="application/x-tex">211</annotation></semantics></math>
we get the following plot.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## create twin relationships</span></span>
<span><span class="va">rel_df2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    id1 <span class="op">=</span> <span class="st">"211"</span>,</span>
<span>    id2 <span class="op">=</span> <span class="st">"212"</span>,</span>
<span>    code <span class="op">=</span> <span class="fl">4</span>,</span>
<span>    famid <span class="op">=</span> <span class="st">"2"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">new_rel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="../reference/Pedigree-class.html">rel</a></span><span class="op">(</span><span class="va">ped2</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">rel_df2</span>, <span class="fu"><a href="../reference/Rel-class.html">Rel</a></span><span class="op">(</span><span class="va">id1</span>, <span class="va">id2</span>, <span class="va">code</span>, <span class="va">famid</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Pedigree-class.html">rel</a></span><span class="op">(</span><span class="va">ped2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/upd_famid.html">upd_famid</a></span><span class="op">(</span><span class="va">new_rel</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">ped2</span><span class="op">)</span></span></code></pre></div>
<p><img src="pedigree_files/figure-html/spouse-1.png" alt="Pedigree with spouse with no children" width="700"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="pedigree-plot-details">Pedigree Plot Details<a class="anchor" aria-label="anchor" href="#pedigree-plot-details"></a>
</h2>
<p>The plot method attempts to adhere to many standards in pedigree
plotting, as presented by <a href="https://pubmed.ncbi.nlm.nih.gov/18792771/" class="external-link">Bennet et al.
2008</a>.</p>
<p>To show some other tricks with pedigree plotting, we use
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mi>e</mi><mi>d</mi><mn>1</mn></mrow><annotation encoding="application/x-tex">ped1</annotation></semantics></math>
from
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>s</mi><mi>a</mi><mi>m</mi><mi>p</mi><mi>l</mi><mi>e</mi><mi>p</mi><mi>e</mi><mi>d</mi></mrow><annotation encoding="application/x-tex">sampleped</annotation></semantics></math>,
which has 41 subjects in 4 generations, including a generation with
double first cousins. After the first marriage of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>114</mn><annotation encoding="application/x-tex">114</annotation></semantics></math>,
they remarried subject
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>113</mn><annotation encoding="application/x-tex">113</annotation></semantics></math>
without children between them. If we do not specify the marriage with
the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>r</mi><mi>e</mi><mi>l</mi><mi>a</mi><mi>t</mi><mi>i</mi><mi>o</mi><mi>n</mi></mrow><annotation encoding="application/x-tex">relation</annotation></semantics></math>
argument, the plot method excludes subject
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>113</mn><annotation encoding="application/x-tex">113</annotation></semantics></math>
from the plot. The basic plot of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mi>e</mi><mi>d</mi><mn>1</mn></mrow><annotation encoding="application/x-tex">ped1</annotation></semantics></math>
is shown in the figure below.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df1</span> <span class="op">&lt;-</span> <span class="va">sampleped</span><span class="op">[</span><span class="va">sampleped</span><span class="op">$</span><span class="va">famid</span> <span class="op">==</span> <span class="fl">1</span>, <span class="op">]</span></span>
<span><span class="va">relate1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    id1 <span class="op">=</span> <span class="fl">113</span>,</span>
<span>    id2 <span class="op">=</span> <span class="fl">114</span>,</span>
<span>    code <span class="op">=</span> <span class="fl">4</span>,</span>
<span>    famid <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span><span class="va">ped1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Pedigree-class.html">Pedigree</a></span><span class="op">(</span><span class="va">df1</span>, <span class="va">relate1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">ped1</span><span class="op">)</span></span></code></pre></div>
<p><img src="pedigree_files/figure-html/plotped1-1.png" alt="Pedigree of family 1" width="700"></p>
<div class="section level3">
<h3 id="align-by-input-order">Align by Input Order<a class="anchor" aria-label="anchor" href="#align-by-input-order"></a>
</h3>
<p>The plot method does a decent job aligning subjects given the order
of the subjects when the Pedigree object is made, and sometimes has to
make two copies of a subject. If we change the order of the subjects
when creating the Pedigree, we can help the plot method reduce the need
to duplicate subjects, as Figure~ no longer has subject
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>110</mn><annotation encoding="application/x-tex">110</annotation></semantics></math>
duplicated.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df1reord</span> <span class="op">&lt;-</span> <span class="va">df1</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">35</span><span class="op">:</span><span class="fl">41</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">34</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">ped1reord</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Pedigree-class.html">Pedigree</a></span><span class="op">(</span><span class="va">df1reord</span>, <span class="va">relate1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">ped1reord</span><span class="op">)</span></span></code></pre></div>
<p><img src="pedigree_files/figure-html/ordering-1.png" alt="Pedigree of family 1 with reordering" width="700"></p>
</div>
<div class="section level3">
<h3 id="plot-colors-and-scales">Plot colors and scales<a class="anchor" aria-label="anchor" href="#plot-colors-and-scales"></a>
</h3>
<p>The <code>Pedigree</code> object contains a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>S</mi><mi>c</mi><mi>a</mi><mi>l</mi><mi>e</mi><mi>s</mi></mrow><annotation encoding="application/x-tex">Scales</annotation></semantics></math>
object that can be modified to change the colors and patterns used in
the plot. To make it easy for the user to modify it a function
<code><a href="../reference/generate_colors.html">generate_colors()</a></code> is available. This function will generate
a color palette for the filling and the bordering of the plot. This
function transform a given column of the dataframe into a factor and
generate a color palette for each level of the factor. The user can then
modify the colors and the patterns used for the filling and the
bordering of the plot.</p>
<p>To do so you can do as follow:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Pedigree-class.html">scales</a></span><span class="op">(</span><span class="va">ped1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class "Scales"</span></span>
<span><span class="co">## Slot "fill":</span></span>
<span><span class="co">##   order column_values    column_mods mods            labels affected  fill density angle</span></span>
<span><span class="co">## 1     1     affection affection_mods    0 Healthy &lt;= to 0.5    FALSE white      NA    NA</span></span>
<span><span class="co">## 2     1     affection affection_mods    1 Affected &gt; to 0.5     TRUE   red      NA    NA</span></span>
<span><span class="co">## 3     1     affection affection_mods   NA              &lt;NA&gt;       NA  grey      NA    NA</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slot "border":</span></span>
<span><span class="co">##   column_values column_mods mods        labels border</span></span>
<span><span class="co">## 1         avail  avail_mods   NA            NA   grey</span></span>
<span><span class="co">## 2         avail  avail_mods    1     Available  green</span></span>
<span><span class="co">## 3         avail  avail_mods    0 Non Available  black</span></span></code></pre>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ped1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generate_colors.html">generate_colors</a></span><span class="op">(</span></span>
<span>    <span class="va">ped1</span>, col_aff <span class="op">=</span> <span class="st">"num"</span>,</span>
<span>    add_to_scale <span class="op">=</span> <span class="cn">TRUE</span>, is_num <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    keep_full_scale <span class="op">=</span> <span class="cn">TRUE</span>, breaks <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    colors_aff <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"green"</span><span class="op">)</span>,</span>
<span>    colors_unaff <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"yellow"</span>, <span class="st">"brown"</span><span class="op">)</span>,</span>
<span>    threshold <span class="op">=</span> <span class="fl">3</span>, sup_thres_aff <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">ped1</span><span class="op">)</span></span></code></pre></div>
<p><img src="pedigree_files/figure-html/generate_colors-1.png" alt="Pedigree of family 1 with change in colors" width="700"></p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># To modify a given scale you can do as follow</span></span>
<span><span class="fu"><a href="../reference/Scales-class.html">fill</a></span><span class="op">(</span><span class="va">ped1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   order column_values    column_mods mods                       labels affected    fill density</span></span>
<span><span class="co">## 1     1     affection affection_mods    0            Healthy &lt;= to 0.5    FALSE   white      NA</span></span>
<span><span class="co">## 2     1     affection affection_mods    1            Affected &gt; to 0.5     TRUE     red      NA</span></span>
<span><span class="co">## 3     1     affection affection_mods   NA                         &lt;NA&gt;       NA    grey      NA</span></span>
<span><span class="co">## 4     2           num       num_mods    1      Affected &lt; to 3 : (1,2]     TRUE #0000FF      NA</span></span>
<span><span class="co">## 5     2           num       num_mods    4    Healthy &gt;= to 3 : [3,4.5]    FALSE #A52A2A      NA</span></span>
<span><span class="co">## 6     2           num       num_mods    3    Healthy &gt;= to 3 : (4.5,6]    FALSE #FFFF00      NA</span></span>
<span><span class="co">## 7     2           num       num_mods    2 Affected &lt; to 3 : [-0.002,1]     TRUE #00FF00      NA</span></span>
<span><span class="co">## 8     2           num       num_mods   NA                      NA : NA       NA    grey      NA</span></span>
<span><span class="co">##   angle</span></span>
<span><span class="co">## 1    NA</span></span>
<span><span class="co">## 2    NA</span></span>
<span><span class="co">## 3    NA</span></span>
<span><span class="co">## 4    NA</span></span>
<span><span class="co">## 5    NA</span></span>
<span><span class="co">## 6    NA</span></span>
<span><span class="co">## 7    NA</span></span>
<span><span class="co">## 8    NA</span></span></code></pre>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Scales-class.html">fill</a></span><span class="op">(</span><span class="va">ped1</span><span class="op">)</span><span class="op">$</span><span class="va">fill</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"#970b6d"</span></span>
<span><span class="fu"><a href="../reference/Scales-class.html">fill</a></span><span class="op">(</span><span class="va">ped1</span><span class="op">)</span><span class="op">$</span><span class="va">density</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">30</span></span>
<span><span class="fu"><a href="../reference/Scales-class.html">fill</a></span><span class="op">(</span><span class="va">ped1</span><span class="op">)</span><span class="op">$</span><span class="va">angle</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">45</span></span>
<span><span class="fu"><a href="../reference/Scales-class.html">border</a></span><span class="op">(</span><span class="va">ped1</span><span class="op">)</span><span class="op">$</span><span class="va">border</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"black"</span>, <span class="st">"orange"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">ped1</span>, legend <span class="op">=</span> <span class="cn">TRUE</span>, leg_loc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">7</span>, <span class="fl">18</span>, <span class="fl">1</span>, <span class="fl">1.5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="pedigree_files/figure-html/generate_colors-2.png" alt="Pedigree of family 1 with change in colors" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="pedigree-utility-functions">Pedigree Utility Functions<a class="anchor" aria-label="anchor" href="#pedigree-utility-functions"></a>
</h2>
<div class="section level3">
<h3 id="ped-as-a-data-frame">Ped as a data.frame<a class="anchor" aria-label="anchor" href="#ped-as-a-data-frame"></a>
</h3>
<p>A main features of a Pedigree object are vectors with an element for
each subject. It is sometimes useful to extract these vectors from the
Pedigree object into a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>d</mi><mi>a</mi><mi>t</mi><mi>a</mi><mi>.</mi><mi>f</mi><mi>r</mi><mi>a</mi><mi>m</mi><mi>e</mi></mrow><annotation encoding="application/x-tex">data.frame</annotation></semantics></math>
with basic information that can be used to construct a new
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mi>e</mi><mi>d</mi><mi>i</mi><mi>g</mi><mi>r</mi><mi>e</mi><mi>e</mi></mrow><annotation encoding="application/x-tex">Pedigree</annotation></semantics></math>
object. This is possible with the <code><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame()</a></code> method,
as shown below.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dfped2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="../reference/Pedigree-class.html">ped</a></span><span class="op">(</span><span class="va">ped2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dfped2</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="subsetting-and-trimming">Subsetting and Trimming<a class="anchor" aria-label="anchor" href="#subsetting-and-trimming"></a>
</h3>
<p>Pedigrees with large size can be a bottleneck for programs that run
calculations on them. The Pedixplorer package contains some routines to
identify which subjects to remove. We show how a subject (e.g. subject
210) can be removed from <sub>ped2</sub>, and how the Pedigree object is
changed by verifying that the <sub>relation</sub> dataframe no longer
has the twin relationship between subjects 210 and 211, as indicated by
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><mi>d</mi><mn>1</mn></mrow><annotation encoding="application/x-tex">id1</annotation></semantics></math>
and <sub>id2</sub>.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ped2_rm210</span> <span class="op">&lt;-</span> <span class="va">ped2</span><span class="op">[</span><span class="op">-</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/Pedigree-class.html">rel</a></span><span class="op">(</span><span class="va">ped2_rm210</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Rel object with 0 relationshipswith 0 MZ twin, 0 DZ twin, 0 UZ twin, 0 Spouse:</span></span>
<span><span class="co">##          id1         id2                     code       famid</span></span>
<span><span class="co">##  &lt;character&gt; &lt;character&gt; &lt;c("ordered", "factor")&gt; &lt;character&gt;</span></span></code></pre>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Pedigree-class.html">rel</a></span><span class="op">(</span><span class="va">ped2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Rel object with 4 relationshipswith 1 MZ twin, 1 DZ twin, 1 UZ twin, 1 Spouse:</span></span>
<span><span class="co">##           id1         id2     code       famid</span></span>
<span><span class="co">##   &lt;character&gt; &lt;character&gt; &lt;factor&gt; &lt;character&gt;</span></span>
<span><span class="co">## 1       2_210       2_211  MZ twin           2</span></span>
<span><span class="co">## 2       2_204       2_208  DZ twin           2</span></span>
<span><span class="co">## 3       2_212       2_213  UZ twin           2</span></span>
<span><span class="co">## 4       2_211       2_212   Spouse           2</span></span></code></pre>
<p>The steps above also works by the <sub>id</sub> of the subjects
themselves.<br>
We provide <em>subset()</em>, which trims subjects from a Pedigree by
their
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><mi>d</mi></mrow><annotation encoding="application/x-tex">id</annotation></semantics></math>
or other argument. Below is an example of removing subject 110, as done
above, then we further trim the Pedigree by a vector of subject ids. We
check the trimming by looking at the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><mi>d</mi></mrow><annotation encoding="application/x-tex">id</annotation></semantics></math>
vector and the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>r</mi><mi>e</mi><mi>l</mi><mi>a</mi><mi>t</mi><mi>i</mi><mi>o</mi><mi>n</mi></mrow><annotation encoding="application/x-tex">relation</annotation></semantics></math>
matrix.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ped2_trim210</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">ped2</span>, <span class="st">"2_210"</span>, keep <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Ped-class.html">id</a></span><span class="op">(</span><span class="fu"><a href="../reference/Pedigree-class.html">ped</a></span><span class="op">(</span><span class="va">ped2_trim210</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "2_201" "2_202" "2_203" "2_204" "2_205" "2_206" "2_207" "2_208" "2_209" "2_211" "2_212" "2_213"</span></span>
<span><span class="co">## [13] "2_214"</span></span></code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Pedigree-class.html">rel</a></span><span class="op">(</span><span class="va">ped2_trim210</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Rel object with 3 relationshipswith 0 MZ twin, 1 DZ twin, 1 UZ twin, 1 Spouse:</span></span>
<span><span class="co">##           id1         id2                     code       famid</span></span>
<span><span class="co">##   &lt;character&gt; &lt;character&gt; &lt;c("ordered", "factor")&gt; &lt;character&gt;</span></span>
<span><span class="co">## 1       2_204       2_208                  DZ twin           2</span></span>
<span><span class="co">## 2       2_212       2_213                  UZ twin           2</span></span>
<span><span class="co">## 3       2_211       2_212                   Spouse           2</span></span></code></pre>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ped2_trim_more</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">ped2_trim210</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"2_212"</span>, <span class="st">"2_214"</span><span class="op">)</span>, keep <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Ped-class.html">id</a></span><span class="op">(</span><span class="fu"><a href="../reference/Pedigree-class.html">ped</a></span><span class="op">(</span><span class="va">ped2_trim_more</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "2_201" "2_202" "2_203" "2_204" "2_205" "2_206" "2_207" "2_208" "2_209" "2_211" "2_213"</span></span></code></pre>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Pedigree-class.html">rel</a></span><span class="op">(</span><span class="va">ped2_trim_more</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Rel object with 1 relationshipwith 0 MZ twin, 1 DZ twin, 0 UZ twin, 0 Spouse:</span></span>
<span><span class="co">##           id1         id2                     code       famid</span></span>
<span><span class="co">##   &lt;character&gt; &lt;character&gt; &lt;c("ordered", "factor")&gt; &lt;character&gt;</span></span>
<span><span class="co">## 1       2_204       2_208                  DZ twin           2</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="shrinking">Shrinking<a class="anchor" aria-label="anchor" href="#shrinking"></a>
</h2>
<p>An additional function in Pedixplorer is <em>shrink()</em>, which
shrinks a Pedigree to a specified bit size while maintaining the maximal
amount of information for genetic linkage and association studies. Using
an indicator for availability and affected status, it removes subjects
in this order: + unavailable with no available descendants + available
and are not parents + available who have missing affected status +
available who are unaffected + available who are affected</p>
<p>We show how to shrink Pedigree 1 to bit size
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>30</mn><annotation encoding="application/x-tex">30</annotation></semantics></math>,
which happens to be the bit size after removing only the unavailable
subjects. We show how to extract the shrunken Pedigree object from the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>s</mi><mi>h</mi><mi>r</mi><mi>i</mi><mi>n</mi><mi>k</mi></mrow><annotation encoding="application/x-tex">shrink</annotation></semantics></math>
result, and plot it.</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">200</span><span class="op">)</span></span>
<span><span class="va">shrink1_b30</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/shrink.html">shrink</a></span><span class="op">(</span><span class="va">ped1</span>, max_bits <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">shrink1_b30</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span><span class="op">:</span><span class="fl">8</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $id_trim</span></span>
<span><span class="co">##  [1] "1_101" "1_102" "1_107" "1_108" "1_111" "1_113" "1_121" "1_122" "1_123" "1_131" "1_132" "1_134"</span></span>
<span><span class="co">## [13] "1_139"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $id_lst</span></span>
<span><span class="co">## $id_lst$unavail</span></span>
<span><span class="co">##  [1] "1_101" "1_102" "1_107" "1_108" "1_111" "1_113" "1_121" "1_122" "1_123" "1_131" "1_132" "1_134"</span></span>
<span><span class="co">## [13] "1_139"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $bit_size</span></span>
<span><span class="co">## [1] 46 29</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $avail</span></span>
<span><span class="co">##  [1] FALSE FALSE FALSE FALSE  TRUE  TRUE FALSE FALSE FALSE  TRUE FALSE  TRUE  TRUE FALSE  TRUE  TRUE</span></span>
<span><span class="co">## [17]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE  TRUE  TRUE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $pedSizeOriginal</span></span>
<span><span class="co">## [1] 41</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $pedSizeIntermed</span></span>
<span><span class="co">## [1] 28</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $pedSizeFinal</span></span>
<span><span class="co">## [1] 28</span></span></code></pre>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">shrink1_b30</span><span class="op">$</span><span class="va">pedObj</span><span class="op">)</span></span></code></pre></div>
<p><img src="pedigree_files/figure-html/shrink1-1.png" alt="Pedigree of family 1 shrinked to 30 bits" width="700"></p>
<p>Now shrink Pedigree 1 to bit size
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>25</mn><annotation encoding="application/x-tex">25</annotation></semantics></math>,
which requires removing subjects who are informative. If there is a tie
between multiple subjects about who to remove, the method randomly
chooses one of them. With this seed setting, the method removes subjects
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>140</mn><annotation encoding="application/x-tex">140</annotation></semantics></math>
then
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>141</mn><annotation encoding="application/x-tex">141</annotation></semantics></math>.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">shrink1_b25</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/shrink.html">shrink</a></span><span class="op">(</span><span class="va">ped1</span>, max_bits <span class="op">=</span> <span class="fl">25</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">shrink1_b25</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span><span class="op">:</span><span class="fl">8</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $id_trim</span></span>
<span><span class="co">##                                                                                                 </span></span>
<span><span class="co">## "1_101" "1_102" "1_107" "1_108" "1_111" "1_113" "1_121" "1_122" "1_123" "1_131" "1_132" "1_134" </span></span>
<span><span class="co">##              id                 </span></span>
<span><span class="co">## "1_139" "1_124" "1_130" "1_125" </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $id_lst</span></span>
<span><span class="co">## $id_lst$unavail</span></span>
<span><span class="co">##  [1] "1_101" "1_102" "1_107" "1_108" "1_111" "1_113" "1_121" "1_122" "1_123" "1_131" "1_132" "1_134"</span></span>
<span><span class="co">## [13] "1_139"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $id_lst$affect</span></span>
<span><span class="co">##      id                 </span></span>
<span><span class="co">## "1_124" "1_130" "1_125" </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $bit_size</span></span>
<span><span class="co">## [1] 46 29 28 26 24</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $avail</span></span>
<span><span class="co">##  [1] FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE  TRUE FALSE  TRUE  TRUE FALSE  TRUE  TRUE  TRUE</span></span>
<span><span class="co">## [17]  TRUE  TRUE FALSE FALSE FALSE FALSE  TRUE  TRUE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $pedSizeOriginal</span></span>
<span><span class="co">## [1] 41</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $pedSizeIntermed</span></span>
<span><span class="co">## [1] 28</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $pedSizeFinal</span></span>
<span><span class="co">## [1] 24</span></span></code></pre>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">shrink1_b25</span><span class="op">$</span><span class="va">pedObj</span><span class="op">)</span></span></code></pre></div>
<p><img src="pedigree_files/figure-html/shrink2-1.png" alt="Pedigree of family 1 shrinked to 25 bits" width="700"></p>
</div>
<div class="section level2">
<h2 id="select-unrelateds">Select Unrelateds<a class="anchor" aria-label="anchor" href="#select-unrelateds"></a>
</h2>
<p>In this section we briefly show how to use
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>u</mi><mi>n</mi><mi>r</mi><mi>e</mi><mi>l</mi><mi>a</mi><mi>t</mi><mi>e</mi><mi>d</mi></mrow><annotation encoding="application/x-tex">unrelated</annotation></semantics></math>
to find a set of the maximum number of unrelated available subjects from
a Pedigree. The input required is a Pedigree object and a vector
indicating availability. In some Pedigrees there are numerous sets of
subjects that satisfy the maximum number of unrelateds, so the method
randomly chooses from the set. We show two sets of subject ids that are
selected by the routine and discuss below.</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ped2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Pedigree-class.html">Pedigree</a></span><span class="op">(</span><span class="va">df2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">set1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/unrelated.html">unrelated</a></span><span class="op">(</span><span class="va">ped2</span><span class="op">)</span></span>
<span><span class="va">set1</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "2_203" "2_206"</span></span></code></pre>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">set2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/unrelated.html">unrelated</a></span><span class="op">(</span><span class="va">ped2</span><span class="op">)</span></span>
<span><span class="va">set2</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "2_203" "2_214"</span></span></code></pre>
<p>We can easily verify the sets selected by
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>u</mi><mi>n</mi><mi>r</mi><mi>e</mi><mi>l</mi><mi>a</mi><mi>t</mi><mi>e</mi><mi>d</mi></mrow><annotation encoding="application/x-tex">unrelated</annotation></semantics></math>
by referring to Figure~ and see that subjects
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>203</mn><annotation encoding="application/x-tex">203</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>206</mn><annotation encoding="application/x-tex">206</annotation></semantics></math>
are unrelated to everyone else in the Pedigree except their children.
Furthermore, we see in df2 that of these two, only subject 203 is
available. Therefore, any set of unrelateds who are available must
include subject 203 and one of the these subjects: 201, 204, 206, 207,
212, and 214, as indicated by the kinship matrix for Pedigree 2 subset
to those with availability status of 1.</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kin2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/kinship.html">kinship</a></span><span class="op">(</span><span class="va">ped2</span><span class="op">)</span></span>
<span><span class="va">is_avail</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Ped-class.html">id</a></span><span class="op">(</span><span class="fu"><a href="../reference/Pedigree-class.html">ped</a></span><span class="op">(</span><span class="va">ped2</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="../reference/Ped-class.html">avail</a></span><span class="op">(</span><span class="fu"><a href="../reference/Pedigree-class.html">ped</a></span><span class="op">(</span><span class="va">ped2</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">kin2</span></span></code></pre></div>
<pre><code><span><span class="co">## 14 x 14 sparse Matrix of class "dsCMatrix"</span></span></code></pre>
<pre><code><span><span class="co">##   [[ suppressing 14 column names '2_201', '2_202', '2_203' ... ]]</span></span></code></pre>
<pre><code><span><span class="co">##                                                                                             </span></span>
<span><span class="co">## 2_201 0.500 .     .    0.250 0.250 0.250 0.250 0.250 .    0.1250 0.1250 0.1250 0.1250 0.1250</span></span>
<span><span class="co">## 2_202 .     0.500 .    0.250 0.250 0.250 0.250 0.250 .    0.1250 0.1250 0.1250 0.1250 0.1250</span></span>
<span><span class="co">## 2_203 .     .     0.50 .     .     .     .     .     .    0.2500 0.2500 .      .      .     </span></span>
<span><span class="co">## 2_204 0.250 0.250 .    0.500 0.250 0.250 0.250 0.250 .    0.2500 0.2500 0.1250 0.1250 0.1250</span></span>
<span><span class="co">## 2_205 0.250 0.250 .    0.250 0.500 0.250 0.250 0.250 .    0.1250 0.1250 0.1250 0.1250 0.1250</span></span>
<span><span class="co">## 2_206 0.250 0.250 .    0.250 0.250 0.500 0.250 0.250 .    0.1250 0.1250 0.1250 0.1250 0.1250</span></span>
<span><span class="co">## 2_207 0.250 0.250 .    0.250 0.250 0.250 0.500 0.250 .    0.1250 0.1250 0.1250 0.1250 0.1250</span></span>
<span><span class="co">## 2_208 0.250 0.250 .    0.250 0.250 0.250 0.250 0.500 .    0.1250 0.1250 0.2500 0.2500 0.2500</span></span>
<span><span class="co">## 2_209 .     .     .    .     .     .     .     .     0.50 .      .      0.2500 0.2500 0.2500</span></span>
<span><span class="co">## 2_210 0.125 0.125 0.25 0.250 0.125 0.125 0.125 0.125 .    0.5000 0.2500 0.0625 0.0625 0.0625</span></span>
<span><span class="co">## 2_211 0.125 0.125 0.25 0.250 0.125 0.125 0.125 0.125 .    0.2500 0.5000 0.0625 0.0625 0.0625</span></span>
<span><span class="co">## 2_212 0.125 0.125 .    0.125 0.125 0.125 0.125 0.250 0.25 0.0625 0.0625 0.5000 0.2500 0.2500</span></span>
<span><span class="co">## 2_213 0.125 0.125 .    0.125 0.125 0.125 0.125 0.250 0.25 0.0625 0.0625 0.2500 0.5000 0.2500</span></span>
<span><span class="co">## 2_214 0.125 0.125 .    0.125 0.125 0.125 0.125 0.250 0.25 0.0625 0.0625 0.2500 0.2500 0.5000</span></span></code></pre>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kin2</span><span class="op">[</span><span class="va">is_avail</span>, <span class="va">is_avail</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## 8 x 8 sparse Matrix of class "dsCMatrix"</span></span>
<span><span class="co">##       2_201 2_203 2_204 2_206 2_207  2_211  2_212  2_214</span></span>
<span><span class="co">## 2_201 0.500  .    0.250 0.250 0.250 0.1250 0.1250 0.1250</span></span>
<span><span class="co">## 2_203 .      0.50 .     .     .     0.2500 .      .     </span></span>
<span><span class="co">## 2_204 0.250  .    0.500 0.250 0.250 0.2500 0.1250 0.1250</span></span>
<span><span class="co">## 2_206 0.250  .    0.250 0.500 0.250 0.1250 0.1250 0.1250</span></span>
<span><span class="co">## 2_207 0.250  .    0.250 0.250 0.500 0.1250 0.1250 0.1250</span></span>
<span><span class="co">## 2_211 0.125  0.25 0.250 0.125 0.125 0.5000 0.0625 0.0625</span></span>
<span><span class="co">## 2_212 0.125  .    0.125 0.125 0.125 0.0625 0.5000 0.2500</span></span>
<span><span class="co">## 2_214 0.125  .    0.125 0.125 0.125 0.0625 0.2500 0.5000</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] Pedixplorer_1.1.3 BiocStyle_2.32.1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] gtable_0.3.5        xfun_0.47           bslib_0.8.0         ggplot2_3.5.1      </span></span>
<span><span class="co">##  [5] htmlwidgets_1.6.4   lattice_0.22-6      quadprog_1.5-8      vctrs_0.6.5        </span></span>
<span><span class="co">##  [9] tools_4.4.1         generics_0.1.3      stats4_4.4.1        tibble_3.2.1       </span></span>
<span><span class="co">## [13] fansi_1.0.6         highr_0.11          pkgconfig_2.0.3     Matrix_1.7-0       </span></span>
<span><span class="co">## [17] data.table_1.16.0   desc_1.4.3          S4Vectors_0.42.1    readxl_1.4.3       </span></span>
<span><span class="co">## [21] lifecycle_1.0.4     stringr_1.5.1       compiler_4.4.1      shinytoastr_2.2.0  </span></span>
<span><span class="co">## [25] textshaping_0.4.0   munsell_0.5.1       httpuv_1.6.15       shinyWidgets_0.8.7 </span></span>
<span><span class="co">## [29] htmltools_0.5.8.1   sass_0.4.9          yaml_2.3.10         lazyeval_0.2.2     </span></span>
<span><span class="co">## [33] plotly_4.10.4       later_1.3.2         pillar_1.9.0        pkgdown_2.1.1      </span></span>
<span><span class="co">## [37] jquerylib_0.1.4     tidyr_1.3.1         DT_0.33             cachem_1.1.0       </span></span>
<span><span class="co">## [41] mime_0.12           tidyselect_1.2.1    digest_0.6.37       stringi_1.8.4      </span></span>
<span><span class="co">## [45] colourpicker_1.3.0  dplyr_1.1.4         purrr_1.0.2         bookdown_0.40      </span></span>
<span><span class="co">## [49] fastmap_1.2.0       grid_4.4.1          colorspace_2.1-1    cli_3.6.3          </span></span>
<span><span class="co">## [53] magrittr_2.0.3      utf8_1.2.4          withr_3.0.1         scales_1.3.0       </span></span>
<span><span class="co">## [57] promises_1.3.0      rmarkdown_2.28      httr_1.4.7          gridExtra_2.3      </span></span>
<span><span class="co">## [61] cellranger_1.1.0    ragg_1.3.3          shiny_1.9.1         evaluate_1.0.0     </span></span>
<span><span class="co">## [65] knitr_1.48          miniUI_0.1.1.1      viridisLite_0.4.2   rlang_1.1.4        </span></span>
<span><span class="co">## [69] Rcpp_1.0.13         xtable_1.8-4        glue_1.8.0          BiocManager_1.30.25</span></span>
<span><span class="co">## [73] BiocGenerics_0.50.0 jsonlite_1.8.9      R6_2.5.1            plyr_1.8.9         </span></span>
<span><span class="co">## [77] systemfonts_1.1.0   fs_1.6.4</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Louis Le Nézet, Jason Sinnwell, Terry Therneau.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
